### DISCUSSION TITLE
How to configure Equinox logging?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-equinox/equinox#20] [Q&A]
The only option I could found so far is `eclipse.log.level` or `log.threshold` mentioned but this configures everything on the global level, but I like to control this on "logger" level (e.g. per bundle), so is there a way using system-properties/config files/ ... or is this a missing feature?

**Comment by HannesWell** [comment]
Good question.
If you just want to trace the SCR, maybe this helps:
https://github.com/apache/felix-dev/tree/master/scr#configuration

**Comment by laeubi** [comment]
> If you just want to trace the SCR, maybe this helps

This will require each and every component to provide custom properties, that's not really useful.

e.g. the slf4j simple logger has the following properties: https://www.slf4j.org/api/org/slf4j/impl/SimpleLogger.html that would be useful here is well.

**Comment by tjwatson** [comment]
External configuration of each `LoggerContext` can be done through the method `org.osgi.service.log.admin.LoggerContext.setLogLevels(Map<String, LogLevel>)` which can be obtained from the `LoggerAdmin` service with the `org.osgi.service.log.admin.LoggerAdmin.getLoggerContext(String)` method.

There is also Configuration Admin which can be used to configure LoggerContext (see https://docs.osgi.org/specification/osgi.core/8.0.0/service.log.html#d0e54441)

We do have built-in configuration property `eclipse.log.level` (set through `config.ini`) that can control the `LogLevel` of the logger named `org.eclipse.equinox.logger`.  This is the logger that writes to the `.log` file and is the logger used by other logging APIs in Eclipse that end up forwarding their messages to the OSGi Logger.  The value can be set to INFO, WARNING and ERROR, INFO will for example show all log message.  I'm open to having a more general property that can somehow be used as a default source for configuring other logger names.  Although I would probably not recommend trying to shoehorn that into the existing `eclipse.log.level` configuration property because that one is awkward and predates the whole OSGi LoggerContext API.

**Comment by laeubi** [comment]
@tjwatson thanks for the insight, I know this could be done through config admin / logger admin, I'm just more searching for something that is used as defaults if no such configuration exits.

For example I currently investigate if we can include [slf4j-osgi](https://github.com/osgi/slf4j-osgi) but this might produce a lot of log outputs, so people will instantly complain about "logspam", so an easy to use (e.g. adding a property to eclipse.ini) would be good, while telling them to fetch OSGi service and doing something in code will most probably be "to complex" :-)

**Reply by tjwatson** [reply]
My intention from my previous comment was to leave an issue open.  I won't have time to look into a solution, but thought it would be useful to have an equinox specific config property that could configure defaults for logging contexts.  This default would be overwritten by config admin or LoggerAdmin if used.

**Reply by laeubi** [reply]
I created https://github.com/eclipse-equinox/equinox/issues/26 to track this!
